File: libsignal/service/src/main/java/org/whispersystems/signalservice/api/push/ServiceIds.java
package org.whispersystems.signalservice.api.push;

import org.whispersystems.signalservice.api.push.ServiceId.ACI;
import org.whispersystems.signalservice.api.push.ServiceId.PNI;

import java.util.Objects;

import okio.ByteString;

/**
 * Helper for dealing with [ServiceId] matching when you only care that either of your
 * service ids match but don't care which one.
 */
public final class ServiceIds {

  private final ACI aci;
  private final PNI pni;

  private ByteString aciByteString;
  private ByteString pniByteString;

  public ServiceIds(ACI aci, PNI pni) {
    this.aci = aci;
    this.pni = pni;
  }

  public ACI getAci() {
    return aci;
  }

  public PNI getPni() {
    return pni;
  }

  public PNI requirePni() {
    return Objects.requireNonNull(pni);
  }

  public boolean matches(ServiceId serviceId) {
    return serviceId.equals(aci) || (pni != null && serviceId.equals(pni));
  }

  public boolean matches(ByteString serviceIdsBytes) {
    if (aciByteString == null) {
      aciByteString = aci.toByteString();
    }

    if (pniByteString == null && pni != null) {
      pniByteString = pni.toByteString();
    }

    return serviceIdsBytes.equals(aciByteString) || serviceIdsBytes.equals(pniByteString);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/keyvalue/ServiceIds.kt
package org.thoughtcrime.securesms.keyvalue

import com.google.protobuf.ByteString
import org.whispersystems.signalservice.api.push.ACI
import org.whispersystems.signalservice.api.push.PNI
import org.whispersystems.signalservice.api.util.UuidUtil
import java.util.UUID

/**
 * Helper for dealing with [ServiceId] matching when you only care that either of your
 * service ids match but don't care which one.
 */
data class ServiceIds(val aci: ACI, val pni: PNI?) {

  private val aciByteString: ByteString by lazy { UuidUtil.toByteString(aci.uuid()) }
  private val pniByteString: ByteString? by lazy { pni?.let { UuidUtil.toByteString(it.uuid()) } }

  fun matches(uuid: UUID): Boolean {
    return uuid == aci.uuid() || uuid == pni?.uuid()
  }

  fun matches(uuid: ByteString): Boolean {
    return uuid == aciByteString || uuid == pniByteString
  }
}
